File: TestLib/src/main/java/com/example/TreesAndGraph/NumberofConnectedComponentinanUndirectedGraph_323.java
package com.example.TreesAndGraph;


import java.nio.charset.CharacterCodingException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class NumberofConnectedComponentinanUndirectedGraph_323 {
//    Input: n = 5 and edges = [[0, 1], [1, 2], [3, 4]]
//
//            0          3
//            |          |
//            1 --- 2    4
//
//    Output: 2
//    https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/
//   time complextity 0(n+m)

    public static TreeNode root = null;
    public static List<List<Integer>> levelOrder = new ArrayList<List<Integer>>();
    public static void main(String[] args) {
//        addNode(3);
        int[][] edges = {{0,1}, {1,2}, {3,4}};
        int result = countComponentsDfs(5, edges);
        result = countComponentsUnionFind(5, edges);
        System.out.print(result);
    }

    //union find path compression & union by size
    public static int countComponentsUnionFind(int n, int[][] edges) {
        int[] parent = new int[n];
        int[] size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
        int components = n;
        for (int[] e : edges) {
            int p1 = find(parent, e[0]);
            int p2 = find(parent, e[1]);
            if (p1 != p2) {
                if (size[p1] < size[p2]) { // Merge small size to large size
                    size[p2] += size[p1];
                    parent[p1] = p2;
                } else {
                    size[p1] += size[p2];
                    parent[p2] = p1;
                }
                components--;
            }
        }
        return components;
    }

    public static int find(int[] roots, int id) {
        while(roots[id] != id) {
            roots[id] = roots[roots[id]];  // optional: path compression
            id = roots[id];
        }
        return id;
    }
    //union find

    //dfs soln
    public static int countComponentsDfs(int nodesCount, int[][] edges) {
        if (nodesCount <= 1) return 1;
        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int i = 0; i < nodesCount; i++) {
            map.put(i, new ArrayList<>());
        }

        for (int[] edge: edges) {
            map.get(edge[0]).add(edge[1]);
            map.get(edge[1]).add(edge[0]);
        }

        int count = 0;
        Set<Integer> visited = new HashSet<>();
        for (int i = 0; i < nodesCount; i++) {
            if (visited.add(i)) {
                dfs(i, visited, map);
                count++;
            }
        }
        return count;
    }

    public static void dfs(int currentNode, Set<Integer> visited, Map<Integer, List<Integer>> map) {
        for (Integer neigh: map.get(currentNode)) {
            if (visited.add(neigh)) {
                dfs(neigh, visited, map);
            }
        }
    }
    //dfs soln










//    public static List<String> countComponentsDfsPrint(List<Character> nodes, char[][] edges) {
//        Map<Character, List<Character>> map = new HashMap<>();
//        List<String> result = new ArrayList<>();
//        for (char[] edge: edges) {
//            map.putIfAbsent(edge[0], new ArrayList<>());
//            map.putIfAbsent(edge[1], new ArrayList<>());
//
//            map.get(edge[0]).add(edge[1]);
//            map.get(edge[1]).add(edge[0]);
//        }
//
//        int count = 0;
//        Set<Character> set = new HashSet<>();
//        for (Character node: nodes) {
//            StringBuilder sb = new StringBuilder();
//            dfsTest(node, map, set, sb);
//            result.add(sb.toString());
//            count++;
//        }
//        return result;
//    }
//
//    public static void dfsTest(Character node, Map<Character, List<Character>> map, Set<Character> set, StringBuilder sb) {
//        if (set.add(node)) {
//            sb.append(node);
//            for (Character nei: map.get(node)) {
//                sb.append("<->");
//                dfsTest(nei, map, set, sb);
//            }
//        }
//    }


    public static void addNode(int value) {
        if (root == null)
         root = new TreeNode(value);
        root.left = new TreeNode(5);
        root.left.left = new TreeNode(6);

        root.left.right = new TreeNode(2);
        root.left.right.left = new TreeNode(7);
        root.left.right.right = new TreeNode(4);

        root.right = new TreeNode(1);
        root.right.right = new TreeNode(8);
        root.right.left = new TreeNode(0);
    }

    // Definition for a Node.
    public static class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode(int x) { val = x; }
  }
}

File: TestLib/src/main/java/com/example/Kotlin/NumberofConnectedComponentinanUndirectedGraph_323.kt
package com.example.Kotlin

import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.HashSet

//    Input: n = 5 and edges = [[0, 1], [1, 2], [3, 4]]
//
//            0          3
//            |          |
//            1 --- 2    4
//
//    Output: 2
//    https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/

fun main (args: Array<String>) {
    val items = arrayOf(intArrayOf(0,1), intArrayOf(1,2), intArrayOf(3,4))
    //println(countComponentsdfs(5,  items))
    println(countComponentsBfs(5,  items))
}

//BFS approach
fun countComponentsBfs(n: Int, edges: Array<IntArray>): Int {
    if (n <= 1) {
        return n;
    }
    var adList: MutableList<MutableList<Int>> = mutableListOf()
    for (i in 0 until n) {
        adList.add(mutableListOf())
    }
    for (i in 0 until edges.size) {
        adList.get(edges[i][0]).add(edges[i][1])
        adList.get(edges[i][1]).add(edges[i][0])
    }

    var count = 0
    var visited = BooleanArray(n){false}
    for (i in 0 until n) {
        if (!visited[i]) {
           count++
            var queue: Deque<Int> = ArrayDeque()
            queue.offer(i)
            while(!queue.isEmpty()) {
                var item = queue.poll()
                visited[item] = true
                var items = adList.get(item)
                for(item in items) {
                    if (!visited[item]) {
                        queue.offer(item)
                    }
                }
            }
        }
    }
    return count
}

//dfs approach
fun countComponentsdfs(n: Int, edges: Array<IntArray>): Int {
     var map:MutableMap<Int, MutableList<Int>> = HashMap()
     for (i in 0 until n) {
         map[i] = ArrayList()
     }

    for (i in 0 until edges.size) {
        map[edges[i][0]]?.add(edges[i][1])
        map[edges[i][1]]?.add(edges[i][0])
    }

    var count = 0
    var visited: MutableSet<Int> = HashSet()
    for (i in 0 until n) {
        if(visited.add(i)) {
            dfs(i, visited, map);
            count++
        }
    }
     return count
}

fun dfs(edge: Int, visited: MutableSet<Int>, map: MutableMap<Int, MutableList<Int>>) {
    var items = map.get(edge)
    for (item in items!!) {
        if(visited.add(item)) {
           dfs(item, visited, map)
        }
    }
}